Adding support for Network Share Hosted Package Source and compatability with generic build tools

nschwerzler opened the pull request:
## Bug
- [Issue 5850](https://github.com/NuGet/Home/issues/5850): Feature Request: NuGet.exe pack should drop resolved nuspec to support network share hosted Package Source
- [Issue 5891](https://github.com/NuGet/Home/issues/5891): Feature Request: Add response file support to NuGet.exe
- [Issue 5893](https://github.com/NuGet/Home/issues/5893): Feature Request: Need the ability for NuGet.exe pack to create a nupkg without the version in the package name
- [Issue 5900](https://github.com/NuGet/Home/issues/5900): Use standard build tool error logging
- [Issue 5892](https://github.com/NuGet/Home/issues/5892): NuGet.exe pack file access is causing warnings in some build tools

Regression: No (New features)

## Fix

- Fixed ability to provide a configuration file for pack command [Issue 5892](https://github.com/NuGet/Home/issues/5892).
- Added build tool friendly error messages for pack commands [Issue 5900](https://github.com/NuGet/Home/issues/5900)
- Added new arg to pack 'InstallPackageToOutputPath' that generates resolved nuspec and sha512 file side by side with nupkg to support share feed. See [Accelerate your network share hosted Package Source](https://blog.nuget.org/20150922/Accelerate-Package-Source.html) for details (Issue 5850). This switch also ommits the version in the package name to help with build tool binplace caching [Issue 5893](https://github.com/NuGet/Home/issues/5893).
- Added response file support for all NuGet.exe input by adding the '@[responseFile]' to allow long command lines [Issue 5891](https://github.com/NuGet/Home/issues/5891). Response files are limited to 2mb to protect from overflows and can be nested up to 3 deep.


## Testing/Validation
Tests Added (Yes): 

NuGet.CommandLine.Test.NuGetPackCommandTest.PackCommand_OutputResolvedNuSpecFileAttemptToOverwriteOriginal
NuGet.CommandLine.Test.NuGetPackCommandTest.PackCommand_InstallPackageToOutputPathWithResponseFile
NuGet.CommandLine.Test.NuGetPackCommandTest.PackCommand_InstallPackageToOutputPath
NuGet.CommandLine.Test.NuGetResponseFileTests.ResponseFileSingleArgFullPath
NuGet.CommandLine.Test.NuGetResponseFileTests.ResponseFileBlankLines
NuGet.CommandLine.Test.NuGetResponseFileTests.NoResponseFile
NuGet.CommandLine.Test.NuGetResponseFileTests.NoArgs
NuGet.CommandLine.Test.NuGetResponseFileTests.ResponseFileTooLarge
NuGet.CommandLine.Test.NuGetResponseFileTests.ResponseFileInvalid
NuGet.CommandLine.Test.NuGetResponseFileTests.ResponseFileMultipleArgs
NuGet.CommandLine.Test.NuGetResponseFileTests.ResponseFileMultiLineArgs
NuGet.CommandLine.Test.NuGetResponseFileTests.ResponseFileMultipleNested
NuGet.CommandLine.Test.NuGetResponseFileTests.ResponseFileInfiniteNestedRecursion
NuGet.CommandLine.Test.NuGetResponseFileTests.ResponseFileFullPath
NuGet.CommandLine.Test.NuGetResponseFileTests.WhiteSpaceArg
NuGet.CommandLine.Test.NuGetResponseFileTests.NullArg
NuGet.CommandLine.Test.NuGetResponseFileTests.ResponseFileLargeArg

Validation done:  

- Executed Build.ps1, all tests passed. 
- Executed all pack tests in Test Explorer and verified passed.
- Ran all new tests and validated passed.

nschwerzler committed:
Adding support for network share hosted Package Source
- Issue 5850: Feature Request: NuGet.exe pack should drop resolved nuspec to support network share hosted Package Source
- Issue 5891: Feature Request: Add response file support to NuGet.exe
- Issue 5893: Feature Request: Need the ability for NuGet.exe pack to create a nupkg without the version in the package name
- Issue 5900: Use standard build tool error logging
- Issue 5892: NuGet.exe pack file access is causing warnings in some build tools

nschwerzler committed:
Removed a couple updates to error strings that were not directly related to pack command

nschwerzler committed:
Updating tests, adding coverage for InstallPackageToOutputPath

dnfclas commented:
@nschwerzler,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

emgarten commented on src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx:
```diff
@@ -6133,4 +6133,19 @@ Oluşturma sırasında NuGet'in paketleri indirmesini önlemek için, Visual Stu
+    <value>error NG5003: Response file '{0}' cannot be empty</value>
+  </data>
+  <data name="Error_ResponseFileTooLarge" xml:space="preserve">
+    <value>error NG5004: Response file '{0}' cannot be larger than {1}mb</value>
```
These codes should be sent through the normal logging interface, not as part of the localized strings. We should also re-evaluate the number ranges used here, and keep it in sync with the rest of the codes. Restore uses `NU` for example so having `NG` here is confusing.

//cc @mishra14

nschwerzler replied:
So you want something like the restore command where it concats the messages to a code and puts all the logging messages thru another method?

nschwerzler replied:
There are multiple loggers used in the project, can you point me to a code example of one so I can use the correct one?

emgarten replied:
Log using `ILogger.Log`
https://github.com/NuGet/NuGet.Client/blob/710ec89011b58a78a2bbff3a88771fbd4606db6b/src/NuGet.Core/NuGet.Common/Logging/ILogger.cs#L32

You can log a `LogMessage`
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/Errors/LogMessage.cs

If you need to extend `LogMessage` into `PackageLogMessage` similar to `RestoreLogMessage`

Logging for nuget.exe to the console happens here: 

https://github.com/NuGet/NuGet.Client/blob/710ec89011b58a78a2bbff3a88771fbd4606db6b/src/NuGet.Clients/NuGet.CommandLine/Common/Console.cs#L384-L406

At that point you can convert the message into the expected format if it differs from the default

nschwerzler replied:
Perfect, thank you.  Will update.

emgarten replied:
You can add the new codes into https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs

I suggest keeping everything in the 5000-6000 range for pack

nschwerzler replied:
Will update codes, I was using the "pack" range defined in the enum comment 1400-1900

nschwerzler replied:
Fixed: Updated logging to use RestoreLogMessage pattern. PackageException now takes codes and pumps the message thru the Console.cs Log(ILogMessage message) method.

emgarten replied:
I still see NG error codes in the resource strings. Please fix these.

emgarten commented on src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs:
```diff
@@ -104,9 +104,17 @@ public void Execute()
                 {
+                    string configFileName = null;
+
+                    PackCommand packCommand = this as PackCommand;
```
Any reason why this is needed? Why is the pack command a special case when other commands are using the config file parameter also?

nschwerzler replied:
I don't think the other commands have config file working to override the default config file. This was the only way I could get config file working for pack and override the default config files. If there is a better way let me know and I can change it.

emgarten commented on src/NuGet.Clients/NuGet.CommandLine/Program.cs:
```diff
@@ -149,15 +154,100 @@ public static int MainCore(string workingDirectory, string[] args)
+        /// </summary>
+        /// <param name="args">The args to parse which can contain response files</param>
+        /// <returns></returns>
+        public static string[] ParseArgsResponseFiles(string[] args)
```
I would expect the response file parsing to be in new helper utility class instead of program.cs

nschwerzler replied:
I will move to a new class

emgarten commented on src/NuGet.Clients/NuGet.CommandLine/Program.cs:
```diff
@@ -334,5 +424,60 @@ private static void SetConsoleInteractivity(IConsole console, Command command)
+        /// <param name="argsLength">Pointer to an int that receives the number of array elements returned</param>
+        /// <returns></returns>
+        [DllImport("shell32.dll", SetLastError = true)]
+        static extern IntPtr CommandLineToArgvW([MarshalAs(UnmanagedType.LPWStr)] string commandLine, out int argsLength);
```
What is the benefit of using this over doing a normal string split? It seems like this will make things difficult for mono/cross platform scenario. I would prefer simple/consistent handling for this across all platforms.

nschwerzler replied:
Nearly all windows command line parsing goes thru this function (including .net). I do not want to reinvent the wheel and have to deal with a lot of unstable code. Here is the documentation reference https://msdn.microsoft.com/en-us/library/windows/desktop/bb776391(v=vs.85).aspx. It does more then just doing a string split. If you know of a better library to use let me know and I can switch to that. I understand your point about mono/cross platform but do not know of a good unified library for that yet. Command line parsing is difficult to get right.

emgarten replied:
Unfortunately mono is a requirement for nuget, I don't think this library is an option.

emgarten replied:
nuget.exe ships as part of mono

emgarten commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,59 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+        private void WriteResolvedNuSpecToPackageOutputDirectory(string packageOutputPath, PackageBuilder packageBuilder)
+        {
+            string packageOutputDirectory = Path.GetDirectoryName(packageOutputPath);
+            string resolvedNuSpecOutputPath = Path.Combine(packageOutputDirectory, packageBuilder.Id + PackagingConstants.ManifestExtension);
```
Use VersionFolderPath resolver to create these names

 https://github.com/NuGet/NuGet.Client/blob/2ea9d7ca853506797b891e736955fe8b47ff1793/src/NuGet.Core/NuGet.Packaging/VersionFolderPathResolver.cs#L171

nschwerzler replied:
Will update

emgarten commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,59 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+                sha512hash = cryptoHashProvider.CalculateHash(fileStream);
+            }
+
+            File.WriteAllText(sha512OutputPath, Convert.ToBase64String(sha512hash));
```
This file should be written to a temp file first, then moved into place using `FileUtility.Move`, this allows clients to safely read a share that has installs/adds going on. The sha512 file is the final marker that lets the client know that a package is safe to read.

nschwerzler replied:
Will update

nschwerzler committed:
Code review feedback
- Using VersionFolderPathResolver for manifest file name
- Moved response file parsing to seperate class

rohit21agrawal commented on src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs:
```diff
@@ -104,9 +104,17 @@ public void Execute()
                 {
+                    string configFileName = null;
+
+                    PackCommand packCommand = this as PackCommand;
```
i don't think we need to make the config file work for pack as it has no use.

have you been able to figure out what's causing the config file access in the pack command? we should fix the real cause

---
In reply to: [143581271](https://github.com/NuGet/NuGet.Client/pull/1735#discussion_r143581271) [](ancestors = 143581271)

rohit21agrawal commented on src/NuGet.Clients/NuGet.CommandLine/Common/CommandLineResponseFile.cs:
```diff
@@ -0,0 +1,154 @@
+        /// <param name="argsLength">Pointer to an int that receives the number of array elements returned</param>
+        /// <returns></returns>
+        [DllImport("shell32.dll", SetLastError = true)]
+        static extern IntPtr CommandLineToArgvW([MarshalAs(UnmanagedType.LPWStr)] string commandLine, out int argsLength);
```
is this going to work on mono? is there a reason we need to use native code to do the parsing? i'd rather use managed code if it's an option at all.

nschwerzler replied:
I can look into reimplementing with managed. I think it is going to be potentially error prone though as that code is a bit complex...

After more review I think we should wrap in preprocessor directive and have different implementations for different operating systems.

nschwerzler replied:
Here is the equivalent for unix: https://linux.die.net/man/3/wordexp. Looks like DllImport using CoreCLR also works on that platform. Calling the low level command line parser is the right thing to do here rather than a re-write in managed.

nschwerzler replied:
ifdef response file support so only desktop is compiled in

nschwerzler replied:
Fixed

emgarten replied:
Desktop is still used for Mono, you will need to check if the OS IsWindows and not Mono I think.

https://github.com/NuGet/NuGet.Client/blob/ceb01f8696f29c8fc2c871180c1e552bd98ac154/src/NuGet.Core/NuGet.Common/RuntimeEnvironmentHelper.cs#L37

nschwerzler replied:
Fixed

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -668,24 +725,36 @@ private PackageArchiveReader BuildFromNuspec(string path)
             PackageBuilder packageBuilder = CreatePackageBuilderFromNuspec(path);
 
-            if (_packArgs.Symbols)
+            PackageArchiveReader package = null;
```
rename to to packageArchiveReader

nschwerzler replied:
Fixed

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -668,24 +725,36 @@ private PackageArchiveReader BuildFromNuspec(string path)
             {
-                // remove source related files when building the lib package
-                ExcludeFilesForLibPackage(packageBuilder.Files);
+                InitCommonPackageBuilderProperties(packageBuilder);
```
move this out to outside if-else

nschwerzler replied:
fixed

rohit21agrawal commented:
>         private PackageArchiveReader BuildFromProjectFile(string path)

you need to add the same if-else condition as you did in BuildFromNuspec method here

---
Refers to: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:783 in f402f13. [](commit_id = f402f132c061e1f7f951d465bf3ec4ecd498d484, deletion_comment = False)

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,67 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+        /// <param name="packageBuilder">The package builder</param>
+        private void WriteResolvedNuSpecToPackageOutputDirectory(string packageOutputPath, PackageBuilder packageBuilder)
+        {
+            string resolvedNuSpecOutputPath = Path.Combine(
```
use the PackCommandRunner.GetOutputhPath function instead

rohit21agrawal replied:
please address this comment

---
In reply to: [143873145](https://github.com/NuGet/NuGet.Client/pull/1735#discussion_r143873145) [](ancestors = 143873145)

rohit21agrawal commented on src/NuGet.Clients/NuGet.CommandLine/Commands/PackCommand.cs:
```diff
@@ -79,8 +79,10 @@ public ICollection<string> Exclude
-        // TODO: Temporarily hide the real ConfigFile parameter from the help text.
-        // When we fix #3230, we should remove this property.
+        [Option(typeof(NuGetCommand), "PackageCommandInstallPackageToOutputPath")]
+        public bool InstallPackageToOutputPath { get; set; }
```
i don't see the OutputFileNamesWithoutVersion switch here.

nschwerzler replied:
Sorry thought we wanted them all combined, will add another switch for the OutputFileNamesWithoutVersion.

nschwerzler replied:
Fixed

nschwerzler committed:
Logging update based on code review feedback

nschwerzler committed:
Variable rename PackageArchiveReader based on code review feedback

emgarten commented on src/NuGet.Core/NuGet.Packaging.Core/PackagingException.cs:
```diff
@@ -2,22 +2,35 @@
-    /// Generic packaging exception.
-    /// </summary>
-    public class PackagingException : Exception
+    public class PackagingException : Exception, ILogMessageException
```
🥇

nschwerzler committed:
Added OutputFileNamesWithoutVersion pack switch based on code review feedback

nschwerzler committed:
Code review feedback

nschwerzler committed:
Add preprocessor directive around windows only code in ResponseFile handling.

rohit21agrawal commented on src/NuGet.Clients/NuGet.CommandLine/Program.cs:
```diff
@@ -46,8 +46,11 @@ public static int Main(string[] args)
-            return MainCore(Directory.GetCurrentDirectory(), args);
+
+            // Find any response files and resolve the args
+            var parsedArgs = CommandLineResponseFile.ParseArgsResponseFiles(args);
```
i would ifdef this out and then ifdef the whole CommandLineResponseFile

nschwerzler committed:
ifdef response file parsing

nschwerzler committed:
Merge branch 'dev-NetworkShareHostedPackageSource' into dev

nschwerzler committed:
Fixing conflict merge

rohit21agrawal commented on src/NuGet.Clients/NuGet.CommandLine/Common/Console.cs:
```diff
@@ -395,7 +395,7 @@ public override void Log(ILogMessage message)
                 else if (message.Level == LogLevel.Error)
                 {
-                    WriteError(message.FormatWithCode());
+                    WriteError("error " + message.FormatWithCode());
```
the error word should  be a string in Resources.resx so it can be localized

nschwerzler replied:
Fixed

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,67 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
                 PrintVerbose(outputPath, builder);
             }
 
+            if (_packArgs.InstallPackageToOutputPath)
```
This is incorrect - you need to do either WriteResolvedNusPecToPackageOutputDirectory + WriteSHA512PackageHash OR builder.Save

your current code will cause the nupkg to be written out too which i believe is not what you want.

nschwerzler replied:
writting the nupkg out is exactly what I want. Here is the desired layout which the code currently does:

microsoft.nuget.packagingexample.nupkg
microsoft.nuget.packagingexample.nupkg.sha512
microsoft.nuget.packagingexample.nuspec

Here is the documentation for the feed layout I am targeting. The only exception is we need to remove the version info from the nupkg and nupkg.sha512 as we will do the step of adding the version back into nupkg and sha512 external to nuget pack using hardlinking to not break build tool caching ([Issue 5893](https://github.com/NuGet/Home/issues/5893))
https://blog.nuget.org/20150922/Accelerate-Package-Source.html

nschwerzler replied:
Does this resolve your comment?

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,67 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+                WriteSHA512PackageHash(outputPath);
+            }
+
             _packArgs.Logger.LogMinimal(String.Format(CultureInfo.CurrentCulture, Strings.Log_PackageCommandSuccess, outputPath));
```
the output path will be the full path to the .nupkg file - you don't want to log that the nupkg was output when you are actually installing it in an extracted manner.

nschwerzler replied:
Not sure I understand this feedback. This is actually our desired logging for the build tools, I want to know where the files are going to be output and if the package creation succeeded.

Here is an example of what the logging looks like now.
Install Package to output path 'd:\packages\microsoft.nuget.packagingexample.nupkg'
Install NuSpec to output path 'd:\packages\microsoft.nuget.packagingexample.nuspec'
Install SHA512 to output path 'd:\packages\microsoft.nuget.packagingexample.nupkg.sha512'
Successfully created package 'd:\packages\microsoft.nuget.packagingexample.nupkg'.

rohit21agrawal reviewed (changes requested):
please address my remaining comments.

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,67 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+        /// <param name="packageBuilder">The package builder</param>
+        private void WriteResolvedNuSpecToPackageOutputDirectory(string packageOutputPath, PackageBuilder packageBuilder)
+        {
+            var resolvedNuSpecOutputPath = Path.Combine(
```
use the GetOutputPath function instead to be consistent

nschwerzler replied:
Fixed

nschwerzler committed:
Code review feedback.
- Moved error string to resource file for localization
- WriteResolvedNuSpecToPackageOutputDirectory and WriteSHA512PackageHash now use GetOutputPath to construct file path
- Added well defined error when input nuspec is not found

nschwerzler commented on src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs:
```diff
@@ -104,9 +104,17 @@ public void Execute()
                 {
+                    string configFileName = null;
+
+                    PackCommand packCommand = this as PackCommand;
```
Sent email with information on details of why need this and/or how you would like it changed.

emgarten commented on src/NuGet.Clients/NuGet.CommandLine/Common/CommandLineResponseFile.cs:
```diff
@@ -0,0 +1,163 @@
+            }
+
+            // Response files are not supported on other platforms yet
+            if (!RuntimeEnvironmentHelper.IsWindows)
```
Add `&& !RuntimeEnvironmentHelper.IsMono`

https://github.com/NuGet/NuGet.Client/blob/ceb01f8696f29c8fc2c871180c1e552bd98ac154/src/NuGet.Core/NuGet.Common/RuntimeEnvironmentHelper.cs#L37

emgarten commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,72 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+            var manifest = new Manifest(new ManifestMetadata(builder), null);
+            using (Stream stream = new FileStream(resolvedNuSpecOutputPath, FileMode.Create))
+            {
+                manifest.Save(stream);
```
Use FileUtility.Replace (doesn't matter if it already exists) to safely write this out to a temp file and then move it

https://github.com/NuGet/NuGet.Client/blob/ceb01f8696f29c8fc2c871180c1e552bd98ac154/src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs#L83

nschwerzler replied:
Fixed

emgarten commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,72 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+            File.WriteAllText(sha512TempOutputPath, Convert.ToBase64String(sha512hash));
+
+            FileUtility.Delete(sha512OutputPath);
+            FileUtility.Move(sha512TempOutputPath, sha512OutputPath);
```
Use FileUtility.Replace, it creates the temp file and does all of this for you.

https://github.com/NuGet/NuGet.Client/blob/ceb01f8696f29c8fc2c871180c1e552bd98ac154/src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs#L83

emgarten commented on src/NuGet.Core/NuGet.Common/Errors/PackLogMessage.cs:
```diff
@@ -0,0 +1,63 @@
+            switch (logLevel)
+            {
+                case LogLevel.Error:
+                    return NuGetLogCode.NU1000;
```
this should be NU5000

emgarten commented on src/NuGet.Core/NuGet.Common/Errors/PackLogMessage.cs:
```diff
@@ -0,0 +1,63 @@
+                case LogLevel.Error:
+                    return NuGetLogCode.NU1000;
+                case LogLevel.Warning:
+                    return NuGetLogCode.NU1500;
```
NU5500 (or whatever the first warning is)

emgarten commented on src/NuGet.Core/NuGet.Packaging/Exceptions/MinClientVersionException.cs:
```diff
@@ -1,15 +1,15 @@
     /// Custom exception type for a package that has a higher minClientVersion than the current client.
     /// </summary>
-    public class MinClientVersionException : PackagingException, ILogMessageException
+    public class MinClientVersionException : Exception, ILogMessageException
```
Why did this change? I'm worried this will break other things. This should stay a PackagingException if possible

nschwerzler replied:
I did this to use the interface instead of making PackagingException a base class. Here is hte error if I change it back. MinClientVersionException.AsLogmessage(): cannot override inherited member 'PackagingException.AsLogMessage(0 because it is not marked virutal abstract or override.

nschwerzler replied:
If you want me to make PackagingException a base class or add another interface IPackagingException here I can do that, or we can leave it like this.

emgarten replied:
I think using `PackagingException` as a base class and making AsLogMessage virtual would be best unless you know of a good reason to avoid this.

nschwerzler committed:
Code review feedback

emgarten commented on src/NuGet.Clients/NuGet.CommandLine/Common/Console.cs:
```diff
@@ -395,7 +395,7 @@ public override void Log(ILogMessage message)
                 else if (message.Level == LogLevel.Error)
                 {
-                    WriteError(message.FormatWithCode());
+                    WriteError(string.Concat(LocalizedResourceManager.GetString("Error"), " ", message.FormatWithCode()));
```
Changing all error messages like this could break other things, is this really needed? If it is I think we would want to make it opt in using some other setting or env var.

nschwerzler replied:
We can isolate the "error" prefix to only be added if it is a Packaging error using the new format. This will not break anything because it is new functionality being added to packaging errors since before packaging errors were just exceptions without special error logging. It is needed because we have a lot of different tools that parse the standard error message format (like what msbuild outputs 'error [code] message') and we do not want to have to change all those.

nschwerzler replied:
Fixed

emgarten commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,72 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+            }
+
+            var manifest = new Manifest(new ManifestMetadata(builder), null);
+            var tempOutputPath = Path.Combine(NuGetEnvironment.GetFolderPath(NuGetFolderPath.Temp), Path.GetFileName(resolvedNuSpecOutputPath));
```
Use: https://github.com/NuGet/NuGet.Client/blob/9d354d48a6d7e85e69e0399c097f7fcf049272b5/src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs#L22

The goal of FileUtility.Replace is atomically update the file, so it needs to be in the same dir for a simple move to make that happen.

emgarten replied:
If you use https://github.com/NuGet/NuGet.Client/blob/9d354d48a6d7e85e69e0399c097f7fcf049272b5/src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs#L83 it will handle writing it to a temp file for you

nschwerzler replied:
I cannot have a temp file with a guid getting dropped into the same directory (var fileName = $"{Guid.NewGuid()}.tmp".ToLowerInvariant();). This would break caching as we have to declare what the output file names are before we run the nuget pack command and any file that shows up during building that is unexpected fails the build. This line in GetTempFilePath would create a temp file name with a guid in the same directory as the package. However, we can get away with creating a temp file name in NuGetEnvironment.GetFolderPath(NuGetFolderPath.Temp) because this folder is a special case in the build system as many tools use it (it is based on Path.GetTempPath()).

nschwerzler committed:
Updating based on code review comments. Making it so error prefix message is only used for packaging errors

nschwerzler committed:
Code review feedback

nschwerzler committed:
Fixing spelling on comment

emgarten commented:
Thanks for the great work and improvements @nschwerzler! 🥇 

I've merge this into dev for 4.6.0 with https://github.com/NuGet/NuGet.Client/commit/f61b406593f3b2a75558acb2056963ee470d7e2b
